File: velocity/src/test/java/dev/rafo/bedrockbridge/velocity/sound/BedrockSoundCatalogTest.java
package dev.rafo.bedrockbridge.velocity.sound;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

class BedrockSoundCatalogTest {
    @TempDir
    Path packs;

    @Test
    void loadsDirectoriesAndArchives() throws Exception {
        Path sounds = Files.createDirectories(packs.resolve("diretorio/sounds"));
        Files.writeString(sounds.resolve("sound_definitions.json"), definitions("lusiadascraft:intro"));

        try (ZipOutputStream archive = new ZipOutputStream(Files.newOutputStream(packs.resolve("outro.mcpack")))) {
            archive.putNextEntry(new ZipEntry("sounds/sound_definitions.json"));
            archive.write(definitions("historia.final").getBytes(java.nio.charset.StandardCharsets.UTF_8));
            archive.closeEntry();
        }

        BedrockSoundCatalog catalog = BedrockSoundCatalog.load(packs, (source, error) -> {});

        assertEquals(2, catalog.size());
        assertEquals("lusiadascraft:intro", catalog.resolve("lusiadascraft:intro"));
        assertEquals("historia.final", catalog.resolve("historia:final"));
    }

    @Test
    void reportsInvalidArchivesWithoutLosingValidSounds() throws Exception {
        Files.writeString(packs.resolve("invalido.mcpack"), "não é um zip");
        Path sounds = Files.createDirectories(packs.resolve("valido/sounds"));
        Files.writeString(sounds.resolve("sound_definitions.json"), definitions("som:valido"));
        List<String> failures = new ArrayList<>();

        BedrockSoundCatalog catalog = BedrockSoundCatalog.load(packs, (source, error) -> failures.add(source));

        assertEquals("som:valido", catalog.resolve("som:valido"));
        assertNull(catalog.resolve("som:ausente"));
        assertTrue(failures.getFirst().endsWith("invalido.mcpack"));
    }

    private static String definitions(String identifier) {
        return "{\"format_version\":\"1.14.0\",\"sound_definitions\":{\""
                + identifier
                + "\":{\"sounds\":[\""
                + identifier
                + "\"]}}}";
    }
}

File: src/test/kotlin/dev/rafo/bedrockbridge/sound/BedrockSoundCatalogTest.kt
package dev.rafo.bedrockbridge.sound

import java.nio.file.Files
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream
import kotlin.io.path.createDirectories
import kotlin.io.path.createTempDirectory
import kotlin.io.path.outputStream
import kotlin.io.path.writeText
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

class BedrockSoundCatalogTest {
    @Test
    fun `lê definições de um pack em diretório`() {
        val packs = createTempDirectory("bedrockbridge-packs")
        val sounds = packs.resolve("Scaffolding/sounds").createDirectories()
        sounds.resolve("sound_definitions.json").writeText(definitions("lusiadascraft:intro"))

        val catalog = BedrockSoundCatalog.load(packs)

        assertEquals("lusiadascraft:intro", catalog.resolve("lusiadascraft:intro"))
        assertEquals(1, catalog.size)
    }

    @Test
    fun `lê definições de um ficheiro mcpack`() {
        val packs = createTempDirectory("bedrockbridge-packs")
        val pack = packs.resolve("scaffolding.mcpack")
        ZipOutputStream(pack.outputStream()).use { archive ->
            archive.putNextEntry(ZipEntry("sounds/sound_definitions.json"))
            archive.write(definitions("story.dialogue.welcome").toByteArray())
            archive.closeEntry()
        }

        val catalog = BedrockSoundCatalog.load(packs)

        assertEquals("story.dialogue.welcome", catalog.resolve("story:dialogue.welcome"))
    }

    @Test
    fun `aceita uma definição sem o namespace minecraft`() {
        val catalog = catalogWith("entity.player.levelup")

        assertEquals("entity.player.levelup", catalog.resolve("minecraft:entity.player.levelup"))
    }

    @Test
    fun `não substitui sons ausentes do pack Bedrock`() {
        val catalog = catalogWith("lusiadascraft:intro")

        assertNull(catalog.resolve("minecraft:block.note_block.pling"))
    }

    @Test
    fun `ignora packs inválidos sem perder os restantes`() {
        val packs = createTempDirectory("bedrockbridge-packs")
        Files.writeString(packs.resolve("invalido.mcpack"), "não é um zip")
        val sounds = packs.resolve("valido/sounds").createDirectories()
        sounds.resolve("sound_definitions.json").writeText(definitions("lusiadascraft:ok"))
        val failures = mutableListOf<String>()

        val catalog = BedrockSoundCatalog.load(packs) { source, _ -> failures += source }

        assertEquals("lusiadascraft:ok", catalog.resolve("lusiadascraft:ok"))
        assertTrue(failures.single().endsWith("invalido.mcpack"))
    }

    @Test
    fun `reconstrói um catálogo sincronizado pelo proxy`() {
        val catalog = BedrockSoundCatalog.fromDefinitions(
            listOf("lusiadascraft:intro", "minecraft:entity.player.levelup"),
        )

        assertEquals(2, catalog.size)
        assertEquals("lusiadascraft:intro", catalog.resolve("lusiadascraft:intro"))
        assertEquals("minecraft:entity.player.levelup", catalog.resolve("minecraft:entity.player.levelup"))
    }

    private fun catalogWith(vararg sounds: String): BedrockSoundCatalog {
        val packs = createTempDirectory("bedrockbridge-packs")
        val directory = packs.resolve("pack/sounds").createDirectories()
        directory.resolve("sound_definitions.json").writeText(definitions(*sounds))
        return BedrockSoundCatalog.load(packs)
    }

    private fun definitions(vararg sounds: String): String {
        val entries = sounds.joinToString(",") { "\"$it\": {\"sounds\": [\"$it\"]}" }
        return """{"format_version":"1.14.0","sound_definitions":{$entries}}"""
    }
}
